### PR TITLE
State management impl

### PR DESCRIPTION
**Pull request by denysLavrienov**
This PR implements deleting files from the user's file list.

Created files:
* `FileListController` - stands for user's file list controller;
* `FileList` - renders file list component;
* `FileListItem` - renders file list item;
* `State` - contains files;
* `Store` - contains current application state and methods to operate it;
* `PubSub` - creates **Event bus** for sharing events between components.

### CODE REVIEW

**Comment by dkalinin** [comment]
@denysLavrienov Please fix conflicts.

**Comment by denysLavrienov** [comment]
@dkalinin PTAL

File: app/component/file-list-item.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
+// import {Button} from './button.js';
```
Comment (dkalinin) [inline app/component/file-list-item.js]: ?

File: app/component/file-list-item.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ *
+ */
```
Comment (dkalinin) [inline app/component/file-list-item.js]: ?

File: app/component/file-list.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import {Component} from './component.js';
+import {FileListItem} from './file-list-item.js';
+
+/**
+ *
+ */
```
Comment (dkalinin) [inline app/component/file-list.js]: ?

File: app/component/router.js [hunk 4]
```diff
@@ -8,9 +8,11 @@ import {Page404} from './page-404.js';
 export class Router {
   /**
    * @param {Object} routingConfig
+   * @param {Store} store
    */
-  constructor(routingConfig) {
+  constructor(routingConfig, store) {
     this.routingConfig = routingConfig;
+    this.store = store;
```
Comment (dkalinin) [inline app/component/router.js]: Why does it need store?

File: app/controller/file-list-controller.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+// import {RequestService} from '../service/request-service.js';
+// import {Router} from '../component/router.js';
```
Comment (dkalinin) [inline app/controller/file-list-controller.js]: ?

File: app/service/pubsub.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    return this.events[event].push(callback);
+  }
+  /**
+   * Executes callback function according to event.
+   * @param {string} event
+   * @param {number} data
+   * @return {number}
+   */
+  publish(event, data) {
+    return this.events[event].map((callback) => callback(data));
```
Comment (dkalinin) [inline app/service/pubsub.js]: this.events[event] can be undefined.

File: app/component/application.js [hunk 7]
```diff
@@ -21,14 +24,56 @@ export class Application {
...
 
-    const router = new Router(routes);
+    const actions = {
+      removeFile(store, id) {
+        store.commit('removeFile', id);
+      },
+      uploadFile(store, file) {
+        store.commit('uploadFile', file);
+      },
+    };
```
Comment (dkalinin) [inline app/component/application.js]: Please move actions and mutations to separate files.

File: app/component/component.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/**
+ * Stands for abstract component, that can render some DOM-object.
+ */
+export class Component {
```
Comment (dkalinin) [inline app/component/component.js]: Please divide a simple Component which can render itself and state aware component which subscribes to store and dispatches events.

File: app/component/file-list-item.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+export class FileListItem {
```
Comment (dkalinin) [inline app/component/file-list-item.js]: Please create an abstract Component and extend it here.

File: app/component/file-list.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+      if (element.type === 'file') {
+        fileListItem = new FileListItem(this.fileList, element.name,
+            element.fileSize, element.uploadDate, element.owner, element.type, element.id);
+      } else if (element.type === 'directory') {
+        const dir = document.createElement('div');
+        fileListItem = new FileListItem(dir, element.name,
+            '', '', element.owner, element.type, element.id);
+        this.fileList.appendChild(dir);
+      }
```
Comment (dkalinin) [inline app/component/file-list.js]: Please refactor FileListItem.

File: app/component/input-field.js [hunk 11]
```diff
@@ -31,42 +33,48 @@ export class InputField {
    */
   set errorField(errorMessage) {
     if (!!errorMessage) {
-      this.label.getElementsByTagName('span')[0].innerHTML = `${errorMessage}<br/>`;
+      this.error.getElementsByTagName('span')[0].innerHTML = `${errorMessage}<br/>`;
```
Comment (dkalinin) [inline app/component/input-field.js]: Let's have a field for span in order to not search for it each time.

File: app/component/input-field.js [hunk 12]
```diff
@@ -31,42 +33,48 @@ export class InputField {
...
   }
 
   /**
    * Creates event `change` for input field.
    * @param {Function} handler function that is executed when field is changed.
+   * @return {Object} this
    */
   onChange(handler) {
-    this.label.addEventListener('change', handler);
+    this.inputWrapper.addEventListener('change', handler);
```
Comment (dkalinin) [inline app/component/input-field.js]: We could add handler to input instead of wrapper.

File: app/model/user-file.js [hunk 13]
```diff
@@ -32,15 +32,15 @@ export class UserFile {
...
+   * Returns file fileSize.
    * @return {string}
    */
   get size() {
     return this._size;
   }
 
   /**
-   * Sets file size value.
+   * Sets file fileSize value.
```
Comment (dkalinin) [inline app/model/user-file.js]: Please revert.

File: app/service/form-validator.js [hunk 14]
```diff
@@ -21,7 +21,7 @@ const IMAGE_FORMATS = ['image/jpeg', 'image/png'];
  */
 export class FormValidator {
   /**
-   * Validates image file by size and file formats.
+   * Validates image file by fileSize and file formats.
```
Comment (dkalinin) [inline app/service/form-validator.js]: Please revert.

File: app/service/form-validator.js [hunk 15]
```diff
@@ -31,7 +31,7 @@ export class FormValidator {
         return 'Wrong file format!';
       }
 
-      if (file.size > IMAGE_SIZE) {
+      if (file.fileSize > IMAGE_SIZE) {
```
Comment (dkalinin) [inline app/service/form-validator.js]: Please revert.

File: app/service/pubsub.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
+/**
+ *
```
Comment (dkalinin) [inline app/service/pubsub.js]: ?

File: app/service/pubsub.js [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+ */
+export default class PubSub {
+  /**
+   *
```
Comment (dkalinin) [inline app/service/pubsub.js]: ?

File: app/component/file-list-item.js [hunk 18]
```diff
@@ -7,21 +7,16 @@ import {Component} from './component.js';
...
-    this.uploadDate = uploadDate;
-    this.owner = owner;
-    this.type = type;
-    this.id = id;
+    this.fileName = file.name;
+    this.fileSize = file.fileSize;
+    this.uploadDate = file.uploadDate;
+    this.owner = file.owner;
+    this.type = file.type;
+    this.id = file.id;
```
Comment (dkalinin) [inline app/component/file-list-item.js]: These fields are not needed or should be private.

File: app/component/input-field.js [hunk 19]
```diff
@@ -52,19 +52,23 @@ export class InputField extends Component {
 
     this.inputWrapper = document.createElement('div');
     this.inputWrapper.className = 'col-xs-8';
-    this.inputWrapper.innerHTML = `
-        <input class="form-control" type="${this.type}" id = "${this.id}" >`;
+    this.input = document.createElement('input');
+    this.input.type = this.type;
+    this.input.id = this.id;
+    this.input.className = 'form-control';
+    this.inputWrapper.appendChild(this.input);
```
Comment (dkalinin) [inline app/component/input-field.js]: inputWrapper field is not needed.

File: app/component/input-field.js [hunk 20]
```diff
@@ -52,19 +52,23 @@ export class InputField extends Component {
 
     this.inputWrapper = document.createElement('div');
     this.inputWrapper.className = 'col-xs-8';
-    this.inputWrapper.innerHTML = `
-        <input class="form-control" type="${this.type}" id = "${this.id}" >`;
+    this.input = document.createElement('input');
```
Comment (dkalinin) [inline app/component/input-field.js]: Make it private and use in value getter.

**Pull request merged**